File: Activations.py
import numpy as np

class Sigmoid:
    def activation(self, Z):
        A = 1 / (1 + np.exp(-Z))
        return A

    def derivation(self, Z):
        return self.activation(Z) * (1 - self.activation(Z))

class ReLU:
    def activation(self, Z):
        A = np.maximum(Z, 0)
        return A

    def derivation(self, Z):
        Z_prime = np.where(Z > 0, 1, 0)
        return Z_prime

class LeakyReLU:
    def __init__(self, negative_slope=0.01):
        self.negative_slope = negative_slope

    def activation(self, Z):
        A = np.maximum(Z, Z*self.negative_slope)
        return A

    def derivation(self, Z):
        Z_prime = np.where(Z > 0, 1, self.negative_slope)
        return Z_prime

class Tanh:
    def activation(self, Z):
        return np.tanh(Z)

    def derivation(self, Z):
        Z_prime = 1 - np.tanh(Z) ** 2
        return Z_prime

class ELU:
    def __init__(self, alpha=1.0):
        self.alpha = alpha

    def activation(self, Z):
        A = np.where(Z > 0, Z, self.alpha * (np.exp(Z) - 1))
        return A

    def derivation(self, Z):
        Z_prime = np.where(Z > 0, 1, self.alpha * np.exp(Z))
        return Z_prime

class Swish:
    def __init__(self, beta=1.0):
        self.beta = beta

    def sigmoid(self, x):
        return 1 / (1 + np.exp(-self.beta * x))

    def activation(self, Z):
        A = Z * self.sigmoid(Z)
        return A

    def derivation(self, Z):
        sigmoid_Z = self.sigmoid(Z)
        Z_prime = sigmoid_Z + Z * sigmoid_Z * (1 - sigmoid_Z) * self.beta
        return Z_prime

File: Activations.cpp
#include <math.h>
#include <vector>
#include "VectorOps.h"
using namespace std;

class Sigmoid{
private:

public:
    vector<vector<double>> activation(const vector<vector<double>>& Z){
        return VectorOps::sigmoid(Z);
    }

    vector<vector<double>> derivation(const vector<vector<double>>& Z){
        vector<vector<double>> sig = VectorOps::sigmoid(Z);
        vector<vector<double>> A(Z.size(), vector<double>(Z[0].size(), 1.0));// all elements 1.0
        return VectorOps::multiply(sig, VectorOps::subtract(A, sig));
    }
};

class ReLU{
private:

public:
    vector<vector<double>> activation(const vector<vector<double>>& Z){
        return VectorOps::max(Z, 0);
    }

    vector<vector<double>> derivation(const vector<vector<double>>& Z){
        vector<vector<double>> Z_prime(Z.size(), vector<double>(Z[0].size()));
        for(size_t i = 0; i < Z.size(); i++){
            for(size_t j = 0; j < Z[0].size(); j++){
                Z_prime[i][j] = Z[i][j] > 0;
            }
        }
        return Z_prime;
    }
};

class LeakyReLU{
private:
    double negative_slope;

public:
    LeakyReLU(double neg_slope=0.01): negative_slope(neg_slope){}

    vector<vector<double>> activation(const vector<vector<double>>& Z){
        return VectorOps::max(Z, VectorOps::multiply(Z, negative_slope));
    }

    vector<vector<double>> derivation(const vector<vector<double>>& Z){
        vector<vector<double>> Z_prime(Z.size(), vector<double>(Z[0].size()));
        for(size_t i = 0; i < Z.size(); i++){
            for(size_t j = 0; j < Z[0].size(); j++){
                Z_prime[i][j] = Z[i][j] > 0 ? 1 : negative_slope;
            }
        }
        return Z_prime;
    }
};

class tanh{
private:

public:
    vector<vector<double>> activation(const vector<vector<double>>& Z){
        return VectorOps::tanh(Z);
    }

    vector<vector<double>> derivation(const vector<vector<double>>& Z){
        vector<vector<double>> A(Z.size(), vector<double>(Z[0].size(), 1.0));// all elements 1.0
        vector<vector<double>> tanh_squared = VectorOps::square(VectorOps::tanh(Z));
        return VectorOps::subtract(A, tanh_squared);
    }
};

class ELU{
private:
    double alpha;

public:
    ELU(double _alpha=1.0):alpha(_alpha){}

    vector<vector<double>> activation(const vector<vector<double>>& Z){
        vector<vector<double>> A(Z.size(), vector<double>(Z[0].size()));
        for(size_t i = 0; i < Z.size(); i++){
            for(size_t j = 0; j < Z[0].size(); j++){
                A[i][j] = Z[i][j] > 0 ? Z[i][j] : alpha * (std::exp(Z[i][j]) - 1);
            }
        }
        return A;
    }

    vector<vector<double>> derivation(const vector<vector<double>>& Z){
        vector<vector<double>> Z_prime(Z.size(), vector<double>(Z[0].size()));
        for(size_t i = 0; i < Z.size(); i++){
            for(size_t j = 0; j < Z[0].size(); j++){
                Z_prime[i][j] = Z[i][j] > 0 ? 1 : alpha * std::exp(Z[i][j]);
            }
        }
        return Z_prime;
    }
};

class Swish{
private:
    double beta;

public:
    Swish(double _beta=1.0):beta(_beta){}

    vector<vector<double>> activation(const vector<vector<double>>& Z){
        return VectorOps::multiply(Z, VectorOps::sigmoid(VectorOps::multiply(Z, beta)));
    }

    vector<vector<double>> derivation(const vector<vector<double>>& Z){
        vector<vector<double>> sigmoid_Z = VectorOps::sigmoid(VectorOps::multiply(Z, beta));
        vector<vector<double>> Z_prime = VectorOps::add(Z_prime, 1);
        Z_prime = VectorOps::multiply(sigmoid_Z, -1);
        Z_prime = VectorOps::multiply(Z_prime, VectorOps::multiply(Z, beta));
        Z_prime = VectorOps::add(Z_prime, 1);
        Z_prime = VectorOps::add(Z_prime, sigmoid_Z);
        return Z_prime;
    }
};